bin/node/test-runner-example/src/lib.rs

```diff
@@ -143,17 +142,10 @@ impl ChainInfo for NodeTemplateChainInfo {
 			backend,
 			keystore.sync_keystore(),
 			task_manager,
-			Box::new(move |_, _| {
-				let slot_duration = slot_duration.clone();
-				async move {
-					let timestamp = sp_timestamp::InherentDataProvider::from_system_time();
-					let slot = sp_consensus_babe::inherents::InherentDataProvider::from_timestamp_and_duration(
-						*timestamp,
-						slot_duration.slot_duration(),
-					);
-
-					Ok((timestamp, slot))
-				}
+			Box::new(|_, client| async move {
+				let provider = SlotTimestampProvider::new(client).map_err(|err| format!("{:?}", err))?;
```

Comment:
Because this closure is a factory function, it needs to be called more than once. This is also evident in the implementation for `CreateInherentDataProviders` trait

https://github.com/paritytech/substrate/blob/90cfb952f2e11bc6d327faf04b30dc5f4cf89df8/primitives/inherents/src/client_side.rs#L43-L60

It's implemented for `Fn`, if a closure captures outer variables, it becomes an `FnOnce` so it can only be called once, because the variables are moved into it once. Luckily whoever implemented the `CreateInherentDataProviders` trait allowed you to pass variables to the factory function, hence the change above.

Reply:
> Because this closure is a factory function, it needs to be called more than once. This is also evident in the implementation for `CreateInherentDataProviders` trait

I know.... But this doesn't mean that you can not do what I said...

https://github.com/paritytech/polkadot/blob/master/node/service/src/lib.rs#L929-L954



> It's implemented for `Fn`, if a closure captures outer variables, it becomes an `FnOnce` so it can only be called once, because the variables are moved into it once. Luckily whoever implemented the `CreateInherentDataProviders` trait allowed you to pass variables to the factory function, hence the change above.

This was not done to pass the client there.